docs/src/variables.md

```diff
@@ -1,23 +1,478 @@
+```jldoctest variables
+julia> @variable(model, y, basename="decision variable")
+decision variable
+```
+This code does four things:
+1. it adds one *optimization* variable to `model`
+2. it creates one *JuMP* variable that acts as a reference to that optimization
+   variable
+3. it binds the JuMP variable to the Julia variable `y`
+4. it tells JuMP that the *name* attribute of this JuMP variable is "decision
+variable". JuMP uses the value of `basename` when it has to print the variable
+as a string.
+
+For example, when we print `y` at the REPL we get:
+```jldoctest variables
+julia> y
+decision variable
+```
+
+Because `y` is a JuMP variable, I can bind it to a different value. For example,
+if I go:
+```jldoctest variables
+julia> y = 1
+1
+```
+`y` is no longer a binding to a JuMP variable. This does not mean that the JuMP
+variable has been destroyed. It still exists and is still a reference to the
+same optimization variable. The binding can be reset by querying the model for
+the symbol *as it was written in the `@variable` macro*. For example:
+```jldoctest variables
+julia> model[:y]
+decision variable
+```
+
+This act of looking up the JuMP variable by using the symbol is most useful when
+composing JuMP models across multiple functions, as illustrated by the following
+example:
+```julia
+function add_component_to_model(model::JuMP.Model)
+    x = model[:x]
+    # ... code that uses x
+end
+function build_model()
+    model = Model()
+    @variable(model, x)
+    add_component_to_model(model)
+end
+# TODO(@odow): add a section on looking up by string
+```
+
+Now that we understand the different between *optimization*, *JuMP*, and *Julia*
+variables, we can properly introduce the `@variable` macro.
 
 The `@variable` macro
 ---------------------
 
-DRAFT: Describe the complete syntax of the `@variable` macro. Anonymous versus
-named variables. Describe the three possible container types returned and how
-to use them (`Array`, `JuMPArray`, and `Dict`).
+We have already seen the basic usage of the `@variable` macro. The next
+extension is to add lower- and upper-bounds to each optimization variable. This
+can be done as follows:
+```jldoctest variables_2; setup=:(model=Model())
+julia> @variable(model, a)
+a
+
+julia> @variable(model, b >= 0)
+b
+
+julia> @variable(model, c <= 1)
+c
+
+julia> @variable(model, 2 <= d <= 3)
+d
+```
+
+!!! note
+    When creating a variable with only a lower-bound or an upper-bound, the name
+    must appear on the left-hand side. Putting the name on the right-hand side
+    (e.g., `@variable(model, 0 <= x)`) will result in an error.
+
+    **Extra for experts:** the reason for this is that at compile time, JuMP
+    does not type and value information. Therefore, the case `@variable(model,
+    a <= b)` is ambiguous as JuMP cannot infer whether `a` is a constant and
+    `b` is the intended variable name, or vice-versa.
+
+We can query whether an optimization variable has a lower- or upper-bound via
```

Comment:
I'm not sure if the discussion about attributes falls entirely under the scope of the `@variable` macro. However, there's a conflict between presenting the common snippets of how to get things done and presenting the complete API for handling attributes (e.g., setting bounds without `@variable`), so maybe we can push this off for now.

Reply:
> setting bounds without @variable

I also don't know how to do this, so that might be a reason it's not in these docs... :)

Reply:
See methods around https://github.com/JuliaOpt/JuMP.jl/blob/133df20e9e5b6fcfd166ac300c9f62110ce4917e/src/variables.jl#L269 :)